zerver/context_processors.py

```diff
@@ -52,6 +53,8 @@ def get_realm_from_request(request: HttpRequest) -> Optional[Realm]:
             request.realm = get_realm(subdomain)
         except Realm.DoesNotExist:
             request.realm = None
+            if raise_exception_if_invalid:
+                raise InvalidSubdomainError()
```

Comment:
This logic is wrong. The first time you call `get_realm_from_request(request, True)`, it will indeed set `request.realm` to `None` and raise the exception. The second time you call `get_realm_from_request(request, True)`, it will see that `hasattr(request, "realm")` is `True` and silently return `None`.

(Another good reason to split this extra functionality into a separate function.)

Reply:
Thanks for catching this! I created a separate function as you said.